ntp.go

```diff
@@ -159,6 +162,87 @@ type QueryOptions struct {
+
+	// (Lambda := RootDelay/2 + RootDispersion) check against MAXDISP (16s)
+	// is required as ntp.org ntpd may report sane other fields while
+	// giving quite erratic clock. The check is declared in packet() at
+	// https://tools.ietf.org/html/rfc5905#appendix-A.5.1.1.
+	lambda := r.RootDelay/2 + r.RootDispersion
+
+	// `r.RTT > 0` check is not included as it does not depend on the
+	// packet itself, but also depends on clock _speed_. It's indicator
+	// that local clock run faster than remote one, so (T4-T1) < (T3-T2),
+	// but it may be local clock issue.
+	// E.g. T1/T2/T3/T4 = 0/10/20/1 leads to RTT = -9s.
+
+	return r.Leap != LeapNotInSync && // RFC5905, packet()
+		0 < r.Stratum && r.Stratum < MaxStratum && // RFC5905, packet()
+		lambda < maxDispersion*time.Second && // RFC5905, packet()
+		!r.Time.Before(r.ReferenceTime) && // RFC5905, packet(), reftime <= xmt ~~ !(xmt < reftime)
+		freshness <= (1<<maxPoll)*time.Second && // ntpdate uses 24h as a heuristics instead of ~36h derived from MAXPOLL
+		ntpEpoch.Before(r.Time) && // sanity
+		ntpEpoch.Before(r.ReferenceTime) // sanity
+}
+
+// Returns single-packet estimate of the root synchronization distance. Some
+// SNTP clients limit check this value before using the response, e.g.
+// systemd-timesyncd uses 5.0s as an upper bound. See also
+// https://tools.ietf.org/html/rfc5905#appendix-A.5.5.2
+func (r *Response) RootDistance() time.Duration {
+	// RFC5905 suggests more strict check against _peer_ in fit(), that
+	// root_dist should be less than MAXDIST + PHI * LOG2D(s.poll).
+	// MAXPOLL is 17, so it is approximately at most (1s + 15e-6 * 2**17) =
+	// 2.96608 s, but MAXDIST and MAXPOLL are confugurable values in the
+	// reference implementation, so only MAXDISP check has hardcoded value
+	// in IsUsable().
+	//
+	// root_dist should also have following summands
+	// + Dispersion towards the peer
+	// + jitter of the link to the peer
+	// + PHI * (current_uptime - peer->uptime_of_last_update)
+	// but all these values are 0 if only single NTP packet was sent.
+	rtt := r.RTT
+	if rtt < 0 {
+		rtt = 0
+	}
+	return (rtt+r.RootDelay)/2 + r.RootDispersion
+}
+
+// Returns amount of causality violation between SNTP timestamps, it may be
+// used as a lower bound of time synchronization error. NB: The "legit" error
+// may be up to 1s during leap second.  SNTP query has four timestamps for
+// consecutive events: T1, T2, T3 and T4. T1 and T4 use local clock, T2 and T3
+// use NTP clock. E.g. this function returns (T1 - T2) if T1 > T2 and
+// everything is OK with T3 / T4 ordering.
+func (r *Response) CausalityViolation() time.Duration {
+	// RTT    = (T4 - T1) - (T3 - T2)     =   T4 - T3 + T2 - T1
+	// Offset = (T2 + T3)/2 - (T4 + T1)/2 = (-T4 + T3 + T2 - T1) / 2
+	// => T2 - T1 = RTT/2 + Offset && T4 - T3 = RTT/2 - Offset
+	// If system wall-clock is synced to NTP-clock then T2 >= T1 && T4 >= T3.
+	// This check is required for chrony as it starts relaying sane NTP
```

Comment:
`chrony` -> `synchrony`

I also feel the comment is a little verbose.  Can you tighten it up, maybe by removing the math?

Reply:
`chrony` is NTP server implementation :)